### PR TITLE
Update juniper

### PR DESCRIPTION
**Pull request by AlexSKaye**
## Description of changes

Updated Juniper compiler from version 2.3.0, previously packaged from prebuilt binaries, to version 4.0.0, now built from source using buildDotnetModule.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by AndersonTorres** [comment]
1. use rebase, not merge
2. run the formatter and merge the result as a separate commit.

**Comment by AlexSKaye** [comment]
Sorry, newbie contributor. Clicked sync on Github, realising now it was totally unnecessary. Hopefully I've fixed things appropriately.

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -951,6 +951,11 @@
     matrix = "@alexshpilkin:matrix.org";
     name = "Alexander Shpilkin";
   };
+  AlexSKaye = {
+    github = "AlexSKaye";
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:955]: An e-mail would be nice.
I suggest you to create a dedicated ProtonMail for GitHub- and other forges- related things.

File: pkgs/development/compilers/juniper/default.nix [hunk 2]
```diff
@@ -1,42 +1,34 @@
...
+  dotnet-runtime = dotnetCorePackages.runtime_8_0;
 
   meta = with lib; {
     description = "Functional reactive programming language for programming Arduino";
-    mainProgram = "juniper";
     longDescription = ''
-      Juniper targets Arduino and supports many features typical of functional programming languages, including algebraic data types, tuples, records,
-      pattern matching, immutable data structures, parametric polymorphic functions, and anonymous functions (lambdas).
-      Some imperative programming concepts are also present in Juniper, such as for, while and do while loops, the ability to mark variables as mutable, and mutable references.
+      The purpose of Juniper is to provide a functional reactive programming platform for designing Arduino projects. FRP's high-level approach to timing-based events fits naturally with Arduino, with which programming almost entirely revolves around reacting to realtime events. Juniper transpiles to Arduino C++, which is then compiled to an Arduino executable.
```
Comment (AndersonTorres) [inline pkgs/development/compilers/juniper/default.nix:26]: Break this lines in 80-char limit

File: pkgs/development/compilers/juniper/default.nix [hunk 3]
```diff
@@ -1,42 +1,34 @@
...
     '';
     homepage = "https://www.juniper-lang.org/";
     license = licenses.mit;
-    maintainers = with maintainers; [ ];
-    platforms = platforms.linux;
+    maintainers = with maintainers; [ AlexSKaye ];
+    mainProgram = "Juniper";
+    platforms = dotnet-sdk.meta.platforms;
```
Comment (AndersonTorres) [inline pkgs/development/compilers/juniper/default.nix:32]:
```suggestion
    inherit (dotnet-sdk.meta) platforms;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AlexSKaye** [comment]
Ugh. Sorry for the commit noise. Looked ok until I pushed. Going to try to clean that up tomorrow when I have a clear head.

File: pkgs/development/compilers/juniper/default.nix [hunk 4]
```diff
@@ -1,42 +1,39 @@
...
-      Some imperative programming concepts are also present in Juniper, such as for, while and do while loops, the ability to mark variables as mutable, and mutable references.
+      The purpose of Juniper is to provide a functional reactive programming
+      platform for designing Arduino projects. FRP's high-level approach to
+      timing-based events fits naturally with Arduino, with which programming
+      almost entirely revolves around reacting to realtime events. Juniper
+      transpiles to Arduino C++, which is then compiled to an Arduino
+      executable.
     '';
     homepage = "https://www.juniper-lang.org/";
     license = licenses.mit;
```
Comment (AndersonTorres) [inline pkgs/development/compilers/juniper/default.nix:34]:
```suggestion
    license = lib.licenses.mit;
```

File: pkgs/development/compilers/juniper/default.nix [hunk 5]
```diff
@@ -1,42 +1,39 @@
...
-    maintainers = with maintainers; [ ];
-    platforms = platforms.linux;
+    maintainers = with maintainers; [ AlexSKaye ];
```
Comment (AndersonTorres) [inline pkgs/development/compilers/juniper/default.nix:35]:
```suggestion
    maintainers = with lib.maintainers; [ AlexSKaye ];
```

File: pkgs/development/compilers/juniper/default.nix [hunk 6]
```diff
@@ -1,42 +1,39 @@
...
-    makeWrapper ${mono}/bin/mono $out/bin/juniper \
-      --add-flags "$out/Juniper.exe \$@"
-    runHook postInstall
-  '';
+  projectFile = "Juniper/Juniper.fsproj";
+  nugetDeps = ./deps.nix;
+  dotnet-sdk = dotnetCorePackages.sdk_8_0;
+  dotnet-runtime = dotnetCorePackages.runtime_8_0;
 
   meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/development/compilers/juniper/default.nix:23]:
```suggestion
  meta = {
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AlexSKaye** [comment]
Right, I think I've integrated all your suggestions and cleaned up the commit history. I read up on Git because my understanding was clearly weak. Hopefully everything is in order. Thanks for your patience.

**Review by AndersonTorres — APPROVED** [review]

**Pull request merged**